### PR TITLE
ENH: ImageInferenceModel to consume more flexible image models

### PR DESCRIPTION
**Pull request by corcra**
Addressing #802.

Description:
I've introduced an abstract parent class ("MinimalImageModel" - open to suggestions for a better title) which the existing `ImageModel` inherits from.
The ImageInferenceEngine then just checks that the image model is of `MinimalImageModel` type.

I'm aware I haven't added any tests - I'm not sure how best to test this. Would it be weird to define a child class of `MinimalImageModel` _inside_ a test and confirm that the ImageInferenceEngine is happy?

### CODE REVIEW

**Comment by fepegar** [comment]
[Converting to draft]

File: hi-ml-multimodal/src/health_multimodal/image/model/model.py [hunk 1]
```diff
@@ -74,7 +75,23 @@ class ImageModelOutput():
...
+class MinimalImageModel(nn.Module):
+    """ An abstract class defining methods required by the ImageInferenceEngine """
+
+    def __init__(self, **kwargs: Any):
+        super().__init__()
+        assert hasattr(self, "training")
+
+    @abstractmethod
+    def forward(self, x: torch.Tensor) -> ImageModelOutput:
+        pass
```
Comment (fepegar) [inline hi-ml-multimodal/src/health_multimodal/image/model/model.py:87]:
```suggestion
        raise NotImplementedError
```

(same for other methods)

**Comment by corcra** [comment]
I'm not sure why the notebook is failing given I haven't touched it 🤔

**Comment by fepegar** [comment]
> I'm not sure why the notebook is failing given I haven't touched it 🤔

It's unrelated to this PR. More info here:
- https://github.com/microsoft/hi-ml/pull/808

Can you pleaser review that PR (do you think that image is appropriate?)? I tried to add you as a reviewer but wasn't able to.

**Comment by fepegar** [comment]
The mysterious notebook error might be related to
- https://github.com/microsoft/hi-ml/issues/725

File: hi-ml-multimodal/src/health_multimodal/image/model/model.py [hunk 2]
```diff
@@ -75,7 +77,23 @@ class ImageModelOutput():
...
-class ImageModel(nn.Module):
+class BaseImageModel(nn.Module, ABC):
+    """ An abstract class defining methods required by the ImageInferenceEngine """
+
+    def __init__(self, **kwargs: Any):
+        super().__init__()
+        assert hasattr(self, "training")
+
+    @abstractmethod
+    def forward(self, x: torch.Tensor) -> ImageModelOutput:
```
Comment (fepegar) [inline hi-ml-multimodal/src/health_multimodal/image/model/model.py:88]: Will this be compatible with a child class taking multiple images?

Also applies to the other method below.

Comment (fepegar) [reply hi-ml-multimodal/src/health_multimodal/image/model/model.py:88]: I guess a solution would be using `*args` and `**kwargs`, although mypy will probably complain. I suppose there are more elegant solutions, though.

File: hi-ml-multimodal/src/health_multimodal/image/model/model.py [hunk 3]
```diff
@@ -75,7 +77,23 @@ class ImageModelOutput():
     projected_patch_embeddings: torch.Tensor
 
 
-class ImageModel(nn.Module):
+class BaseImageModel(nn.Module, ABC):
+    """ An abstract class defining methods required by the ImageInferenceEngine """
+
+    def __init__(self, **kwargs: Any):
+        super().__init__()
+        assert hasattr(self, "training")
```
Comment (fepegar) [inline hi-ml-multimodal/src/health_multimodal/image/model/model.py:85]: Why is this necessary?

File: hi-ml-multimodal/src/health_multimodal/image/model/model.py [hunk 4]
```diff
@@ -75,7 +77,23 @@ class ImageModelOutput():
     projected_patch_embeddings: torch.Tensor
 
 
-class ImageModel(nn.Module):
+class BaseImageModel(nn.Module, ABC):
+    """ An abstract class defining methods required by the ImageInferenceEngine """
```
Comment (fepegar) [inline hi-ml-multimodal/src/health_multimodal/image/model/model.py:81]:
```suggestion
    """Abstract class for image models."""
```

File: hi-ml-multimodal/src/health_multimodal/image/model/model.py (hunk 3 above)
Comment (fepegar) [reply hi-ml-multimodal/src/health_multimodal/image/model/model.py:85]: I suspect this `__init__` method can just be removed completely, but I might be wrong.

File: hi-ml-multimodal/src/health_multimodal/image/model/model.py (hunk 2 above)
Comment (corcra) [reply hi-ml-multimodal/src/health_multimodal/image/model/model.py]: I've pushed a possible solution where we use an `ImageModelInput` class, a child class that takes >1 image could then subclass this to add new inputs. This is more explicit than using `*args/**kwargs` but it feels a bit clunky.

File: hi-ml-multimodal/src/health_multimodal/image/model/model.py (hunk 3 above)
Comment (corcra) [reply hi-ml-multimodal/src/health_multimodal/image/model/model.py:90]: I put it in initially because we use the `training` attribute in the inference engine, but I now realise we can assume it has one by virtue of being a `nn.module`, so lets remove this `__init__`!

File: hi-ml-multimodal/src/health_multimodal/image/model/model.py (hunk 2 above)
Comment (fepegar) [reply hi-ml-multimodal/src/health_multimodal/image/model/model.py]: How about just removing the method? Then children will have freedom to implement their `forward`s however they want.

**Comment by fepegar** [comment]
> I'm not sure why the notebook is failing given I haven't touched it 🤔

FYI: https://github.com/microsoft/hi-ml/issues/725#issuecomment-1448184728

File: hi-ml-multimodal/src/health_multimodal/image/model/model.py (hunk 2 above)
Comment (corcra) [reply hi-ml-multimodal/src/health_multimodal/image/model/model.py]: If the forward doesn't return something with at least some of the fields in `ImageModelOutput`, our inference engine methods will fail later.

Comment (fepegar) [reply hi-ml-multimodal/src/health_multimodal/image/model/model.py]: Sure, but the children classes could take care of that.

Comment (fepegar) [reply hi-ml-multimodal/src/health_multimodal/image/model/model.py]: Anyway, this is not terribly important and could be changed in a future PR.

Comment (fepegar) [reply hi-ml-multimodal/src/health_multimodal/image/model/model.py]: Actually, I've changed my mind. Looking at the code, this breaks the current API of the model, which is not great, and passing some custom class to the model instead of just a tensor is not very friendly or PyTorch-standard. So I think it might be better to use the *args syntax or simply delegate the creation of the `forward` method to the children classes.

File: hi-ml-multimodal/src/health_multimodal/image/inference_engine.py [hunk 5]
```diff
@@ -78,7 +79,7 @@ def get_projected_global_embedding(self, image_path: Path) -> torch.Tensor:
                  where joint_feature_dim is the dimensionality of the joint latent space.
         """
         input_image, _ = self.load_and_transform_input_image(image_path, self.transform)
-        projected_img_emb = self.model.forward(input_image).projected_global_embedding
+        projected_img_emb = self.model.forward(ImageModelInput(image=input_image)).projected_global_embedding
```
Comment (Shruthi42) [inline hi-ml-multimodal/src/health_multimodal/image/inference_engine.py:82]: Could we break this up into multiple lines? Creating the image model input first then doing the forward pass, etc.

File: hi-ml-multimodal/src/health_multimodal/image/model/model.py [hunk 6]
```diff
@@ -139,7 +158,7 @@ def get_patchwise_projected_embeddings(self, input_img: torch.Tensor, normalize:
         :returns projected_embeddings: tensor of embeddings in shape [batch, n_patches_h, n_patches_w, feature_size].
         """
         assert not self.training, "This function is only implemented for evaluation mode"
-        outputs = self.forward(input_img)
+        outputs = self.forward(ImageModelInput(image=input_img))
```
Comment (Shruthi42) [inline hi-ml-multimodal/src/health_multimodal/image/model/model.py:161]: Similar to the comment above, let's split this up into different lines.

File: hi-ml-multimodal/src/health_multimodal/image/model/__init__.py [hunk 7]
```diff
@@ -3,13 +3,15 @@
...
 
+from .model import BaseImageModel
 from .model import ImageModel
 from .model import ResnetType
 from .model import get_biovil_resnet
 from .model import CXR_BERT_COMMIT_TAG
 from .model import BIOMED_VLP_CXR_BERT_SPECIALIZED
 
 __all__ = [
+    "BaseImageModel",
```
Comment (Shruthi42) [inline hi-ml-multimodal/src/health_multimodal/image/model/__init__.py:14]: We might also want to add `ImageModelInput` here, any code that subclasses `BaseImageModel` will need to access this class.

File: hi-ml-multimodal/src/health_multimodal/image/inference_engine.py (hunk 5 above)
Comment (fepegar) [reply hi-ml-multimodal/src/health_multimodal/image/inference_engine.py:82]: I suggest keeping the current API so we don't break it: https://github.com/microsoft/hi-ml/pull/806#discussion_r1124704433

File: hi-ml-multimodal/src/health_multimodal/image/model/model.py (hunk 2 above)
Comment (fepegar) [reply hi-ml-multimodal/src/health_multimodal/image/model/model.py]: Sounds good!

File: hi-ml-multimodal/src/health_multimodal/image/model/__init__.py (hunk 7 above)
Comment (fepegar) [reply hi-ml-multimodal/src/health_multimodal/image/model/__init__.py:14]: [Resolving as we've changed method and this is now outdated]

File: hi-ml-multimodal/src/health_multimodal/image/model/model.py [hunk 8]
```diff
@@ -112,7 +125,9 @@ def train(self, mode: bool = True) -> Any:
             self.projector.train(mode=False)
         return self
 
-    def forward(self, x: torch.Tensor) -> ImageModelOutput:
+    def forward(self, *args: Any, **kwargs: Any) -> ImageModelOutput:
+        x = kwargs["x"]
+        assert isinstance(x, torch.Tensor)
```
Comment (fepegar) [inline hi-ml-multimodal/src/health_multimodal/image/model/model.py:130]: Why do we need these changes?

File: hi-ml-multimodal/src/health_multimodal/image/inference_engine.py (hunk 5 above)
Comment (fepegar) [reply hi-ml-multimodal/src/health_multimodal/image/inference_engine.py]: [Resolving as it's now outdated]

File: hi-ml-multimodal/src/health_multimodal/image/model/model.py (hunk 6 above)
Comment (fepegar) [reply hi-ml-multimodal/src/health_multimodal/image/model/model.py]: [Resolving as it's now outdated]

File: hi-ml-multimodal/src/health_multimodal/image/model/model.py (hunk 8 above)
Comment (fepegar) [reply hi-ml-multimodal/src/health_multimodal/image/model/model.py:130]: [Discussed offline]

Comment (corcra) [reply hi-ml-multimodal/src/health_multimodal/image/model/model.py:130]: Discussed offline - we don't, let's simplify :)

File: hi-ml-multimodal/src/health_multimodal/image/model/model.py [hunk 9]
```diff
@@ -125,9 +125,7 @@ def train(self, mode: bool = True) -> Any:
             self.projector.train(mode=False)
         return self
 
-    def forward(self, *args: Any, **kwargs: Any) -> ImageModelOutput:
-        x = kwargs["x"]
-        assert isinstance(x, torch.Tensor)
+    def forward(self, x: torch.Tensor, *args: Any, **kwargs: Any) -> ImageModelOutput:  # type: ignore
```
Comment (fepegar) [inline hi-ml-multimodal/src/health_multimodal/image/model/model.py:128]: Is the `x` to enforce that at least one tensor is in the `args`?

Comment (fepegar) [reply hi-ml-multimodal/src/health_multimodal/image/model/model.py:128]: Could you please add the specific mypy error that is being ignored? E.g. `# type: ignore[arg-type]` (not sure about the exact error code you're getting, which I guess is the point of adding it. It's safer than a blanket `# type: ignore`.

**Review by fepegar — CHANGES_REQUESTED** [review]

**Review by fepegar — APPROVED** [review]

File: hi-ml-multimodal/src/health_multimodal/image/model/model.py [hunk 10]
```diff
@@ -112,7 +125,7 @@ def train(self, mode: bool = True) -> Any:
             self.projector.train(mode=False)
         return self
 
-    def forward(self, x: torch.Tensor) -> ImageModelOutput:
+    def forward(self, x: torch.Tensor, *args: Any, **kwargs: Any) -> ImageModelOutput:  # type: ignore[override]
```
Comment (Shruthi42) [inline hi-ml-multimodal/src/health_multimodal/image/model/model.py:128]: Why are we allowing and then ignoring potential args and kwargs here? If someone passes in two tensors to this function, there's no indication that this is not supported - it's very possible this fails silently completely unknown to the user.

We are designing this code to allow users to use classes that inherit from `BaseImageModel` interchangeably, so there's a reasonable chance that a user swaps out one class with another and the model forward pass is run incorrectly but silently.

Comment (fepegar) [reply hi-ml-multimodal/src/health_multimodal/image/model/model.py:128]: I agree. I thought we were ignoring mypy errors for the base class, not for this.

Comment (corcra) [reply hi-ml-multimodal/src/health_multimodal/image/model/model.py]: I've removed the unused args/kwargs now. The mypy issue occurs in the child class:
`model.py:128: error: Signature of "forward" incompatible with supertype "BaseImageModel"` 
... so I think we need to ignore it here. This might be a bit annoying as I guess all children of `BaseImageModel` will need to do something similar to appease mypy.

**Review by Shruthi42 — APPROVED** [review]

**Comment by fepegar** [comment]
It looks like this PR will be superseded by
- #830 

which has diverged anyway. Shall we close this / put it on hold?

**Comment by corcra** [comment]
> It looks like this PR will be superseded by
> 
> * [ENH: Add BioViL-T image and text encoders #830](https://github.com/microsoft/hi-ml/pull/830)
> 
> which has diverged anyway. Shall we close this / put it on hold?

Yes, sounds good. I'll close this in favour of #830.
